Aijuh/199/log level selector

Aijuh opened the pull request:
Finish implementing log level selector as @kangz12345 requested as a week goal.
Doubly checked comments and grammer issue. 
I will fix code right away if I missed something.

This closes #199

Aijuh committed:
Add log level selector

1. Import Qcombobox (line 10)
2. Add Qcombobox in loggerFrame
3. make initLevelSelector function to initialize combobox
4. make fixLevel function to respond to combobox's change

Aijuh committed:
Follow pylint guidline

Aijuh committed:
Add a punctuation(.).

kangz12345 commented on examples/logger.py:
```diff
@@ -8,6 +8,7 @@
 
 from PyQt5.QtCore import QObject, pyqtSlot, pyqtSignal
 from PyQt5.QtWidgets import QWidget, QVBoxLayout, QPushButton, QTextEdit, QLabel, QDialogButtonBox
+from PyQt5.QtWidgets import QComboBox
```
If the import line becomes too long, you can use parenthesis:
```python
from module import (
    Class1, Class2, Class3,
    Class4, Class5, Class6, Class7,
)
```

BECATRUE replied:
Please merge this line into the above one!

If the line is too long, write it down as follows:
```python
from PyQt5.QtWidgets import (
    QWidget, QVBoxLayout, ...
)
```

BECATRUE commented on examples/logger.py:
```diff
@@ -63,10 +64,12 @@ def __init__(self, parent: Optional[QObject] = None):
         self.logEdit = QTextEdit(self)
         self.logEdit.setReadOnly(True)
         self.clearButton = QPushButton("Clear")
+        self.levelSelector = QComboBox(self)
```
Please add this widget into the attributes docstring!

BECATRUE replied:
Oh, and please rename it to include 'Box`.
For legibilty, we always include the type in each widget's name.

kangz12345 commented on examples/logger.py:
```diff
@@ -63,10 +64,12 @@ def __init__(self, parent: Optional[QObject] = None):
         self.logEdit = QTextEdit(self)
         self.logEdit.setReadOnly(True)
         self.clearButton = QPushButton("Clear")
+        self.levelSelector = QComboBox(self)
```
We use the type of the widget as its variable name.
For example, `clearButton` is a push button, `logEdit` is a text edit.
Therefore, `levelBox` would be a valid candidate (you can use any other names that ends with box).

We follow this policy because it helps recognize the widget class instantaneously, and it describes the widget quite well.

BECATRUE commented on examples/logger.py:
```diff
@@ -133,6 +136,29 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        self.initLevelSelector(self.loggerFrame.levelSelector)
+
+    def initLevelSelector(self, box: QComboBox):
+        """Initializes a combo box for selecting level."""
```
Please add the docstring about `box`.

kangz12345 replied:
You should write a argument docstring:

```python
"""Initializes ... level.

Args:
    box: ...
"""
```

BECATRUE commented on examples/logger.py:
```diff
@@ -133,6 +136,29 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+
+    def fixLevel(self, text):
+        """Responds to the levelSelector widget and changes the handler's level."""
+        if text == "DEBUG":
```
I suggest a simpler solution!

```python
self.handler.setLevel(getattr(logging, text))
```

Aijuh replied:
I used this method at first, but after reading @kangz12345 's suggestion I used dictionary instead.

kangz12345 commented on examples/logger.py:
```diff
@@ -133,6 +136,29 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+
+    def initLevelSelector(self, box: QComboBox):
+        """Initializes a combo box for selecting level."""
+        box.addItem("DEBUG")
+        box.addItem("INFO")
+        box.addItem("WARNING")
+        box.addItem("ERROR")
+        box.addItem("CRITICAL")
```
You can use [`addItems()`](https://doc.qt.io/qt-5/qcombobox.html#addItems):
```python
box.addItems(("DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"))
```

BECATRUE commented on examples/logger.py:
```diff
@@ -133,6 +136,29 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        box.activated[str].connect(self.fixLevel)
+
+    def fixLevel(self, text):
+        """Responds to the levelSelector widget and changes the handler's level."""
```
Please add the docstring and type hint about `text`.

kangz12345 replied:
Please write the argument docstring.

BECATRUE commented on examples/logger.py:
```diff
@@ -133,6 +136,29 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+
+    def initLevelSelector(self, box: QComboBox):
+        """Initializes a combo box for selecting level."""
+        box.addItem("DEBUG")
```
Referred to [addItems()](https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QComboBox.html#PySide2.QtWidgets.PySide2.QtWidgets.QComboBox.addItems), you can add all items in just a line!

kangz12345 commented on examples/logger.py:
```diff
@@ -133,6 +136,29 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        box.addItem("WARNING")
+        box.addItem("ERROR")
+        box.addItem("CRITICAL")
+        box.activated[str].connect(self.fixLevel)
```
Could you explain about the `activated[str]`? Is there any reference about this?
I found [`textActivated`](https://doc.qt.io/qt-5/qcombobox.html#textActivated).

Aijuh replied:
I read [activated](https://doc.qt.io/qt-5/qcombobox.html#activated).
The signal 'activated[str]' emits signal as string when comboBox's current text is changed.  
I think textActivated is more appropriate. I will use it.

kangz12345 replied:
However, its argument is `int` (`index`). I wonder where you found the usage of the `[str]`.

kangz12345 commented on examples/logger.py:
```diff
@@ -133,6 +136,29 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        box.addItem("CRITICAL")
+        box.activated[str].connect(self.fixLevel)
+
+    def fixLevel(self, text):
```
Please annotate the type of `text`.

BECATRUE commented on examples/logger.py:
```diff
@@ -133,6 +136,29 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        box.addItem("WARNING")
+        box.addItem("ERROR")
+        box.addItem("CRITICAL")
+        box.activated[str].connect(self.fixLevel)
```
Referred to [activated](https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QComboBox.html#PySide2.QtWidgets.PySide2.QtWidgets.QComboBox.activated), this signal is deprecated.

Please use another function!
I suggest a solution.

```python
box.activated.connect(self.fixLevel)

def fixLevel(self):
    text = self.loggerFrame.levelSelector.currentText()
    ...
```

BECATRUE replied:
Also please move this line into the constructor!
We usually connect a signal to a slot in there.

Aijuh replied:
I am not sure whether signal 'activated' has problem, so I changed to 'currentTextChanged'. It emits signal whenever currentText is changed.
I moved codes to construtor.

Aijuh replied:
I used `textActivated` instead, not using `activated`.
I moved it to constructor.

kangz12345 commented on examples/logger.py:
```diff
@@ -133,6 +136,29 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+
+    def fixLevel(self, text):
+        """Responds to the levelSelector widget and changes the handler's level."""
+        if text == "DEBUG":
+            self.handler.setLevel(logging.DEBUG)
+        if text == "INFO":
+            self.handler.setLevel(logging.INFO)
+        if text == "WARNING":
+            self.handler.setLevel(logging.WARNING)
+        if text == "ERROR":
+            self.handler.setLevel(logging.ERROR)
+        if text == "CRITICAL":
+            self.handler.setLevel(logging.CRITICAL)
```
1. `text` can be any string and `fixLevel()` is a public method. In this case, you should notify that "nothing happens when an invalid `text` is given". Or, if you can guarantee that invalid inputs are never given, you can change it to an internal (private) method `_fixLevel()`.
2. Use `elif` for efficiency.

In this case, the code is simple enough so might not be a problem, but there are several other approaches for this kind of task (you may keep the current code):
- Use [`enum`](https://docs.python.org/3/library/enum.html)
- Use [`getattr()`](https://docs.python.org/3/library/functions.html#getattr)
- Use a dictionary:
```python
level = {
    "DEBUG": logging.DEBUG,
    "INFO": logging.INFO,
    ...
}[text]
```

Aijuh replied:
I will use dictionary and set function as public.

BECATRUE reviewed (changes requested):
I left some reviews!

kangz12345 commented on examples/logger.py:
```diff
@@ -133,6 +136,29 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        box.addItem("CRITICAL")
+        box.activated[str].connect(self.fixLevel)
+
+    def fixLevel(self, text):
```
The name `fixLevel` can be interpreted as it fixes (freezes) the level as well.
How about `setLevel`?

Aijuh replied:
That seems reasonable. I will use it.

kangz12345 commented:
Could you attach the screenshot of the GUI?

Aijuh committed:
Update feedback

Aijuh committed:
Update logger.py

Aijuh committed:
Update logger.py

Aijuh commented:
Updated codes by reflecting feedback.
1. Removed the initialization function for setLevelBox and move its code to the class constructor.
2. Follow comment guideline and naming convention.
3. Change `activate[str]` to `textActivated`.
4. Change nested if-else structure to dictionary.

Following is screenshot of the frame
![캡처](https://github.com/snu-quiqcl/qiwis/assets/43592775/425d30f8-492e-417c-a88d-676f2ed8f0f7)

BECATRUE commented on examples/logger.py:
```diff
@@ -54,6 +56,7 @@ class LoggerFrame(QWidget):
     Attributes:
         logEdit: A textEdit which shows all logs.
         clearButton: A button for clearing all logs.
+        setLevelBox: A combo box for setting the logger's level.
```
It's a minor suggestion, but how about changing 'combo box' into 'comboBox' similar to `logEdit`?

BECATRUE commented on examples/logger.py:
```diff
@@ -133,6 +138,24 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        self.loggerFrame.setLevelBox.textActivated.connect(self.setLevel)
+
+    def setLevel(self, text: str):
+        """Responds to the levelSelector widget and changes the handler's level.
```
Please change 'levelSelector' into 'setLevelBox'!

BECATRUE commented on examples/logger.py:
```diff
@@ -133,6 +138,24 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+            "INFO": logging.INFO,
+            "WARNING":logging.WARNING,
+            "ERROR":logging.ERROR,
+            "CRITICAL":logging.CRITICAL
```
Please add a space after `:`.

BECATRUE reviewed (changes requested):
Probably this may be a fianl review!

Aijuh committed:
change comment

kangz12345 commented on examples/logger.py:
```diff
@@ -63,10 +66,12 @@ def __init__(self, parent: Optional[QObject] = None):
         self.logEdit = QTextEdit(self)
         self.logEdit.setReadOnly(True)
         self.clearButton = QPushButton("Clear")
+        self.setLevelBox = QComboBox(self)
```
`setLevelBox` seems like a method name which sets the levelBox.
Of course it is for _setting_ the level, but it also _shows_ the current level.
I think `levelBox` describes its role well enough.

kangz12345 commented on examples/logger.py:
```diff
@@ -133,6 +138,24 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        self.loggerFrame.setLevelBox.addItems(["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"])
+        self.loggerFrame.setLevelBox.textActivated.connect(self.setLevel)
+
+    def setLevel(self, text: str):
```
How about adding the `pyqtSlot(str)` decorator?

Aijuh committed:
Reflect feedback from @kangz12345

kangz12345 commented on examples/logger.py:
```diff
@@ -54,6 +56,7 @@ class LoggerFrame(QWidget):
     Attributes:
         logEdit: A textEdit which shows all logs.
         clearButton: A button for clearing all logs.
+        setLevelBox: A comboBox for setting the logger's level.
```
Update this to `levelBox`!

Aijuh replied:
Sorry. I should have double checked.

kangz12345 commented on examples/logger.py:
```diff
@@ -133,6 +138,25 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        """Responds to the setLevelBox widget and changes the handler's level.
+        
+        Args:
+            text: Selected level in the level select box.
```
Please add explanation about the input validity. For example:
`It should be one of DEBUG, INFO, WARNING, ERROR and CRITICAL (case-sensitive). Any other input is ignored.`

kangz12345 reviewed (changes requested):
Almost there!

Aijuh committed:
update comments

Aijuh committed:
Update logger.py

kangz12345 commented on examples/logger.py:
```diff
@@ -133,6 +138,27 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        Args:
+            text: Selected level in the level select box.
+                  It should be one of 'DEBUG', 'INFO', 'WARNING', 'ERROR' and 'CRITICAL'.
+                  it should be case-sensitive and any other input is ignored.
```
Please start with a capital letter: `It should be ...`.

kangz12345 commented on examples/logger.py:
```diff
@@ -133,6 +138,27 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+
+        Args:
+            text: Selected level in the level select box.
+                  It should be one of 'DEBUG', 'INFO', 'WARNING', 'ERROR' and 'CRITICAL'.
```
Not a big deal but please use double quotes `"` instead of `'`.

Aijuh committed:
Mistake fixed

Aijuh commented:
Fixed mistakes.

BECATRUE reviewed (approved):
LGTM
